include/boost/numeric/ublas/tensor/detail/extents_functions.hpp

```diff
@@ -0,0 +1,376 @@
+      }
+    }
+  }
+
+  template <class T, T... E>
+  inline
+  constexpr auto squeeze_impl( basic_static_extents<T,E...> const& e ){
+    
+    using extents_type = basic_static_extents<T,E...>;
+
+    if constexpr( extents_type::size() <= typename extents_type::size_type(2) ){
+      return e;
+    }
+
+    using value_type = typename extents_type::value_type;
+    using size_type = typename extents_type::size_type;
+
+    auto one_free_static_extents = squeeze_impl_remove_one(e);
+
+    // check after removing 1s from the list are they same
+    // if same that means 1s does not exist and no need to
+    // squeeze
+    if constexpr( decltype(one_free_static_extents)::Rank != extents_type::Rank ){
+      
+      // after squeezing, all the extents are 1s we need to
+      // return extents of (1, 1)
+      if constexpr( decltype(one_free_static_extents)::Rank == size_type(0) ){
+
+        return basic_static_extents<T, value_type(1),value_type(1)>{};
+
+      }else if constexpr( decltype(one_free_static_extents)::Rank == (1) ){
+        // to comply with GNU Octave this check is made
+        // if position 2 contains 1 we push at back
+        // else we push at front
+        if constexpr( extents_type::at(1) == value_type(1) ){
+          return decltype( push_back<T, value_type(1)>(one_free_static_extents) ){};
+        }else{
+          return decltype( push_front<T, value_type(1)>(one_free_static_extents) ){};
+        }
+
+      }else{
+        return one_free_static_extents;
+      }
+
+    }else{
+      return e;
+    }
+    
+  }
+
+  template <class T>
+  inline
+  constexpr auto squeeze_impl( basic_extents<T> const& e ){
+    using extents_type  = basic_extents<T>;
+    using base_type     = typename extents_type::base_type;
+    using value_type    = typename extents_type::value_type;
+    using size_type    = typename extents_type::size_type;
+    
+    if( e.size() <= size_type(2) ){
+      return e;
+    }
+
+    auto not_one = [](auto const& el){
+      return el != value_type(1);
+    };
+
+    // count non one values
+    size_type size = std::count_if(e.begin(), e.end(), not_one);
+    
+    // reserve space
+    base_type n_extents( std::max(size, size_type(2)), 1 );
+    
+    // copying non 1s to the new extents
+    std::copy_if(e.begin(), e.end(), n_extents.begin(), not_one);
+
+    // checking if extents size goes blow 2
+    // if size of extents goes to 1
+    // complying with GNU Octave
+    // if position 2 contains 1 we 
+    // swap the pos
+    if( size < size_type(2) && e[1] != value_type(1) ){
+      std::swap(n_extents[0], n_extents[1]);
+    }
+      
+    return extents_type(n_extents); 
+  }
+
+  template <class T, size_t N>
+  inline
+  constexpr auto squeeze_impl( basic_fixed_rank_extents<T,N> const& e ){
+    if constexpr( N <= 2 ){
+      return e;
+    }else{
+      return squeeze_impl(e.to_dynamic_extents());
+    }
+  }
```

Comment:
This should be placed inside the corresponding header of the extents definitions.

Reply:
these are one-time use for a squeeze that's it or helper functions. So there is no need to separate them.